src/LaunchDarkly.EventSource/AsyncHelpers.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace LaunchDarkly.EventSource
+{
+    internal static class AsyncHelpers
+    {
+        // General-purpose timeout logic that uses a timed CancellationToken instead of
+        // Task.Delay. The cancellationToken that is passed in is for explicitly cancelling
+        // the operation from elsewhere; DoWithTimeout wraps this to create a token that can
+        // also be cancelled by a timeout, and passes that token to taskFn.
```

Comment:
This replaces the previous logic where, in order to do an async read with a timeout, we would create an additional task that _is_ the timeout (with `Task.Delay`) and then use `Task.WhenAny` to see which one of those two tasks (the read and the timeout) finished first.

.NET has a better way to introduce timeouts into async tasks: create a `CancellationToken` with a `TimeSpan` parameter, which causes it to fire a cancellation after that amount of time. We already have another `CancellationToken` that carries the "we are explicitly cancelling because they called either Close or Restart" signal, so we wire them together with `CreateLinkedTokenSource`. Now we have a single cancel signal, and if it goes off, we can tell whether it was a timeout by checking whether the original "explicitly cancel" signal was used or not.

This is still slightly trickier than it sounds because .NET's support for cancellation tokens is inconsistent - see below.